### PR TITLE
#3659 #3660 WPF/GTK ListView ContextMenu, SelectionBehavior fix

### LINKED ISSUE #3659: WPF ListView Cell ContextMenu
Create listview with custom template
For ViewCell create contextmenu.

AR
Right click works the same as left click

ER
Display contextmenu

Works on Andoid, iOS, macOS. 
Version tested 3.1 and higher.

### LINKED ISSUE #3660: WPF ListView Cell Selection Behaviour
Create ListView and add a few items. Add Event to command behavior on ItemTapped.
Click Tab until you get focus on the first item.
Click Down arrow on keyboard.

AR
The second item is selected, ItemTapped event is raised

ER
The second item is selected.

XF version 3.1 and higher

### PR DESCRIPTION
**Pull request by VladislavAntonyuk**
### Issues Resolved ### 

- fixes #3659
- fixes #3660

### API Changes ###
 
 None

### Platforms Affected ### 

- WPF
- GTK
### Behavioral/Visual Changes ###

None

### Before/After Screenshots ### 

Not applicable

### PR Checklist ###

- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

File: Xamarin.Forms.Platform.WPF/Renderers/ListViewRenderer.cs [hunk 1]
```diff
@@ -156,10 +156,20 @@ void UpdateHorizontalScrollBarVisibility()
 		}
 
 		void OnNativeKeyUp(object sender, KeyEventArgs e)
-			=> Element.NotifyRowTapped(Control.SelectedIndex, cell: null);
+		{
+			if (e.Key == Key.Enter)
```
Comment (VladislavAntonyuk) [inline Xamarin.Forms.Platform.WPF/Renderers/ListViewRenderer.cs]: #3660

**Comment by rmarinho** [comment]
Build is failing you can check here https://dev.azure.com/xamarin/public/_build/results?buildId=2728&view=results

File: Xamarin.Forms.Core/ListView.cs [hunk 2]
```diff
@@ -421,7 +421,7 @@ public string GetDisplayTextFromGroup(object cell)
 		}
 
 		[EditorBrowsable(EditorBrowsableState.Never)]
-		public void NotifyRowTapped(int groupIndex, int inGroupIndex, Cell cell = null)
+		public void NotifyRowTapped(int groupIndex, int inGroupIndex, Cell cell = null, bool isContextMenuRequested = false)
```
Comment (samhouts) [inline Xamarin.Forms.Core/ListView.cs]: I think this is a breaking change. We need to make another overload that keeps the same signature. Right, @PureWeen ?

File: Xamarin.Forms.Core/ListView.cs [hunk 3]
```diff
@@ -440,23 +440,28 @@ public void NotifyRowTapped(int groupIndex, int inGroupIndex, Cell cell = null)
...
-			ItemTapped?.Invoke(this, new ItemTappedEventArgs(ItemsSource.Cast<object>().ElementAt(groupIndex), cell?.BindingContext, TemplatedItems.GetGlobalIndexOfItem(cell?.BindingContext)));
+				ItemTapped?.Invoke(this,
+					new ItemTappedEventArgs(ItemsSource.Cast<object>().ElementAt(groupIndex), cell?.BindingContext,
+						TemplatedItems.GetGlobalIndexOfItem(cell?.BindingContext)));
+			}
 		}
 
 		[EditorBrowsable(EditorBrowsableState.Never)]
-		public void NotifyRowTapped(int index, Cell cell = null)
+		public void NotifyRowTapped(int index, Cell cell = null, bool isContextmenuRequested = false)
```
Comment (samhouts) [inline Xamarin.Forms.Core/ListView.cs]: same as above

File: Xamarin.Forms.Core/ListView.cs (hunk 2 above)
Comment (VladislavAntonyuk) [reply Xamarin.Forms.Core/ListView.cs]: @samhouts, am I understand you correct, that we'll have 4 methods?
void NotifyRowTapped(int index, int inGroupIndex, Cell cell, bool isContextMenuRequested); 
void NotifyRowTapped(int index, Cell cell, bool isContextMenuRequested); 
void NotifyRowTapped(int index, Cell cell);  
void NotifyRowTapped(int index, int inGroupIndex, Cell cell); 

For now code is not changed for execution if you not set isContextMenuRequested.

Comment (StephaneDelcroix) [reply Xamarin.Forms.Core/ListView.cs]: @samhouts right, this breaks the ABI

File: Xamarin.Forms.Core/IListViewController.cs [hunk 4]
```diff
@@ -13,8 +13,8 @@ public interface IListViewController : IViewController
 
 		Cell CreateDefaultCell(object item);
 		string GetDisplayTextFromGroup(object cell);
-		void NotifyRowTapped(int index, int inGroupIndex, Cell cell);
-		void NotifyRowTapped(int index, Cell cell);
+		void NotifyRowTapped(int index, int inGroupIndex, Cell cell, bool isContextMenuRequested = false);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/IListViewController.cs]: you can't change a public interface

**Comment by jsuarezruiz** [comment]
There is also a small conflict in the GTK `NavigationPageRenderer`.

**Comment by VladislavAntonyuk** [comment]
@jsuarezruiz , @StephaneDelcroix any other changes requested?

**Comment by VladislavAntonyuk** [comment]
@jsuarezruiz Rebased on 4.3.0. Added Issue6417

File: Xamarin.Forms.Core/IListViewController.cs [hunk 5]
```diff
@@ -15,6 +15,8 @@ public interface IListViewController : IViewController
 		string GetDisplayTextFromGroup(object cell);
 		void NotifyRowTapped(int index, int inGroupIndex, Cell cell);
 		void NotifyRowTapped(int index, Cell cell);
+		void NotifyRowTapped(int index, int inGroupIndex, Cell cell, bool isContextMenuRequested);
```
Comment (samhouts) [inline Xamarin.Forms.Core/IListViewController.cs]: Adding to a public interface is also a breaking change. We'd need to add a new interface.

Comment (VladislavAntonyuk) [reply Xamarin.Forms.Core/IListViewController.cs]: ListView : IListViewController, IListViewController2

Public interface IListViewController2
{
void NotifyRowTapped(int index, int inGroupIndex, Cell cell, bool isContextMenuRequested);
}

Something like this?

File: Xamarin.Forms.Core/ListView.cs [hunk 6]
```diff
@@ -10,7 +9,7 @@
 namespace Xamarin.Forms
 {
 	[RenderWith(typeof(_ListViewRenderer))]
-	public class ListView : ItemsView<Cell>, IListViewController, IElementConfiguration<ListView>
+	public class ListView : ItemsView<Cell>, IListViewWithContextMenuController, IElementConfiguration<ListView>
```
Comment (samhouts) [inline Xamarin.Forms.Core/ListView.cs]:
```suggestion
	public class ListView : ItemsView<Cell>, IListViewController, IElementConfiguration<ListView>
```

File: Xamarin.Forms.Core/IListViewController.cs [hunk 7]
```diff
@@ -2,6 +2,12 @@
...
 {
+	public interface IListViewWithContextMenuController : IListViewController
+	{
+		void NotifyRowTapped(int index, int inGroupIndex, Cell cell, bool isContextMenuRequested);
+		void NotifyRowTapped(int index, Cell cell, bool isContextMenuRequested);
+	}
+
 	public interface IListViewController : IViewController
 	{
 		event EventHandler<ScrollToRequestedEventArgs> ScrollToRequested;
```
Comment (samhouts) [inline Xamarin.Forms.Core/IListViewController.cs]:
```suggestion
		event EventHandler<ScrollToRequestedEventArgs> ScrollToRequested;
		void NotifyRowTapped(int index, int inGroupIndex, Cell cell, bool isContextMenuRequested);
		void NotifyRowTapped(int index, Cell cell, bool isContextMenuRequested);
```

File: Xamarin.Forms.Core/IListViewController.cs [hunk 8]
```diff
@@ -2,6 +2,12 @@
 
 namespace Xamarin.Forms
 {
+	public interface IListViewWithContextMenuController : IListViewController
```
Comment (samhouts) [inline Xamarin.Forms.Core/IListViewController.cs]:
```suggestion
```

File: Xamarin.Forms.Core/IListViewController.cs [hunk 9]
```diff
@@ -2,6 +2,12 @@
...
+	{
```
Comment (samhouts) [inline Xamarin.Forms.Core/IListViewController.cs]:
```suggestion
```

File: Xamarin.Forms.Core/IListViewController.cs [hunk 10]
```diff
@@ -2,6 +2,12 @@
...
+		void NotifyRowTapped(int index, int inGroupIndex, Cell cell, bool isContextMenuRequested);
```
Comment (samhouts) [inline Xamarin.Forms.Core/IListViewController.cs]:
```suggestion
```

File: Xamarin.Forms.Core/IListViewController.cs [hunk 11]
```diff
@@ -2,6 +2,12 @@
...
+		void NotifyRowTapped(int index, Cell cell, bool isContextMenuRequested);
```
Comment (samhouts) [inline Xamarin.Forms.Core/IListViewController.cs]:
```suggestion
```

File: Xamarin.Forms.Core/IListViewController.cs [hunk 12]
```diff
@@ -2,6 +2,12 @@
...
+	}
```
Comment (samhouts) [inline Xamarin.Forms.Core/IListViewController.cs]:
```suggestion
```

File: Xamarin.Forms.Core/IListViewController.cs [hunk 13]
```diff
@@ -2,6 +2,12 @@
...
+
```
Comment (samhouts) [inline Xamarin.Forms.Core/IListViewController.cs]:
```suggestion
```

**Comment by VladislavAntonyuk** [comment]
@samhouts, we violate Interface Segregation by extending interface with method, which not all child need to use, don't we? Also we can break down clients, which uses IListViewController in there own projects.

**Comment by samhouts** [comment]
@VladislavAntonyuk Only `ListView` uses `IListViewController`. We expect that users will extend the `*Renderer` classes and would be very unlikely to use the `*Controller` classes on their own.

**Comment by rmarinho** [comment]
Do you mind rebase so we can fix the build?

**Comment by VladislavAntonyuk** [comment]
@rmarinho rebased on master

**Comment by VladislavAntonyuk** [comment]
@samhouts, friendly reminder. any plans to merge this PR?

**Comment by VladislavAntonyuk** [comment]
@rmarinho please review

**Comment by VladislavAntonyuk** [comment]
@samhouts what is the target branch I should use?

**Comment by samhouts** [comment]
@VladislavAntonyuk 5.0.0, please! Thank you!

**Comment by VladislavAntonyuk** [comment]
@samhouts done

**Pull request merged**